ingenialink/ethercat/network.py

```diff
@@ -455,56 +446,96 @@ def load_firmware(
-                subprocess.run(
-                    f"chmod 777 {exec_path}",
-                    check=True,
-                    shell=True,
-                    encoding="utf-8",
-                )
-            except subprocess.CalledProcessError as e:
-                raise ILFirmwareLoadError("Could not change the FoE binary permissions.") from e
-        try:
-            if sys_name == "linux":
-                subprocess.run(
-                    f"{exec_path} {self.interface_name} {slave_id} {fw_file} "
-                    f"{int(boot_in_app)} {password}",
-                    check=True,
-                    shell=True,
-                    encoding="utf-8",
-                )
+        self._check_slave_id(slave_id)
+
+        slave = self._ecat_master.slaves[slave_id - 1]
+        if not boot_in_app:
+            self._force_boot_mode(slave)
+        self._switch_to_boot_state(slave)
+
+        foe_result = self._write_foe(slave, fw_file, password)
+
+        if foe_result < 0:
+            error_message = "The firmware file could not be loaded correctly."
+            if foe_result in self.__FOE_ERRORS:
+                error_message += f" {self.__FOE_ERRORS[foe_result]}."
             else:
-                subprocess.run(
-                    [
-                        exec_path,
-                        self.interface_name,
-                        f"{slave_id}",
-                        fw_file,
-                        f"{int(boot_in_app)}",
-                        f"{password}",
-                    ],
-                    check=True,
-                    shell=True,
-                    encoding="utf-8",
-                )
-        except subprocess.CalledProcessError as e:
-            foe_return_error = self.FOE_ERRORS.get(e.returncode, self.UNKNOWN_FOE_ERROR)
-            raise ILFirmwareLoadError(
-                f"The firmware file could not be loaded correctly. {foe_return_error}"
-            ) from e
+                error_message += f" Error code: {foe_result}."
+            raise ILFirmwareLoadError(error_message)
+        self.__init_nodes()
         logger.info("Firmware updated successfully")
 
+    def _check_slave_id(self, slave_id: int) -> None:
+        """Check if the slave ID is valid and that it is detected.
+
+        Args:
+            slave_id: The slave ID.
+
+        """
+        if not isinstance(slave_id, int) or slave_id < 0:
+            raise ValueError("Invalid slave ID value")
+        if not self.__is_master_running:
+            self._start_master()
+        if slave_id not in self.__last_init_nodes:
+            self.__init_nodes()
+        if len(self.__last_init_nodes) == 0:
+            raise ILError("Could not find any slaves in the network.")
+        if slave_id not in self.__last_init_nodes:
+            raise ILError(f"Slave {slave_id} was not found.")
+
+    def _switch_to_boot_state(self, slave: "CdefSlave") -> None:
+        """Request the transition to the boot state."""
```

Comment:
"Request" typically indicates that it works asynchronously.

I would go with: Transitions the slave to boot state.

Reply:
Done